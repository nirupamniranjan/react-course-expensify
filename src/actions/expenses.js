
import uuid from 'uuid'
import database from '../firebase/firebase'
import { add } from 'numeral';
export const addExpense = (expense) =>({
    type: 'ADD_EXPENSE',
    expense 

});

export const startAddExpense = (expenseData={description :'',
note:'',
amount:0,
createdAt:0})=>{

    return (dispatch)=>{
        const {description ='',
        note='',
        amount=0,
        createdAt=0}=expenseData;
        const expense={description,note,amount,createdAt};

        
       const submitData=database.ref('expenses').push(expenseData);
        
       return submitData.then((ref)=>{
          
            dispatch(addExpense({
                id:ref.key,
                ...expense
            }))
        }).catch((e)=>{
            console.log(e + 'error occured')
        })
    }
}
export const removeExpense=({id} ={})=>({
    type: 'REMOVE_EXPENSE',
     id 
});

//EDIT Expense

export const editExpense =(id,updates) =>({
    type:'EDIT_EXPENSE',
    id,
    updates
});

export const setExpenses = (expenses) =>({
    type : 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = ()=>{

    return (dispatch)=>{
        
      return database.ref('expenses').once('value').then((snapshot)=>{
       const expenses = [];
       snapshot.forEach((childSnapshot)=>{

        expenses.push({
            id : childSnapshot.key,
            ...childSnapshot.val()
       })
    })
        
        dispatch(setExpenses(expenses))
    
    
    })
        
       
    }
}
