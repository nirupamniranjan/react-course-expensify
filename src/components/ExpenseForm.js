import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'
import 'react-dates/initialize'



const date= moment();

class ExpenseForm extends React.Component{

    constructor(props){
        super(props);
        this.state ={
            description: props.expense?props.expense.description:'',
            textArea:props.expense?props.expense.textArea:'',
            amount:props.expense?(props.expense.amount/100).toString():'',
            createdAt:props.expense? moment(props.expense.createdAt): moment(),
            CalanderOnfocus:false,
            error:''
         };
        
    }
   onDiscrptionChange=(e)=>{
       const description = e.target.value;
       this.setState(()=>({description}));
   }
  
   onTextAreaChange = (e)=>{
       const textArea =e.target.value;
       this.setState(()=>({textArea}));
   }
   OnAmountChange =(e)=>{
       const amount =e.target.value;

       if(!amount||amount.match(/^\d{1,}(\.\d{0,2})?$/)){
           this.setState(({amount}));
       }

   };
   onDateChange=(createdAt)=>{
       if(createdAt){
    this.setState(()=>({createdAt}))
   }
}
onSubmit=(e)=>{
    e.preventDefault();

    if(!this.state.description||!this.state.amount){
        this.setState(()=> ({error:'Please provide description and amount.'}));
    }
    else{
        this.setState(()=>({error:''}));
        this.props.onSubmit({
            description: this.state.description,
            amount:parseFloat(this.state.amount,10),
            createdAt:this.state.createdAt.valueOf(),
            textArea:this.state.textArea
        })
    }
}

   onFocusChange=({focused})=>{
       this.setState(()=>({CalanderOnfocus:focused}));
   }
    render(){
        return (
            <div>
            {this.state.error && <p>{this.state.error}</p>}
            <form onSubmit={this.onSubmit}>

           <input type="text" 
           placeholder="Description" 
           autoFocus 
           value={this.state.description}
           onChange={this.onDiscrptionChange}
            />
           
           <input type="text" placeholder="amount"
            value={this.state.amount}
            onChange={this.OnAmountChange}
           />
           <SingleDatePicker
           date={this.state.createdAt}
           onDateChange={this.onDateChange}
           focused={this.state.CalanderOnfocus}
           onFocusChange={this.onFocusChange}
           numberOfMonths={1}
           isOutsideRange={(day)=>false}
           />
           <textarea
            placeholder="Add a note for your expense(Optiona)"
            value={this.state.textArea}
            onChange={this.onTextAreaChange}
          >


           </textarea>
            <button>Add Expense</button>
            </form>
        
               

            </div>
        );
    }
}

export default ExpenseForm;