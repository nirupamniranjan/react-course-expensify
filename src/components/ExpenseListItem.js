import React from 'react'
import {Link} from 'react-router-dom'


const expensetListItem=(props)=>(


    <div>
        <p>
            Description:{props.expense.description} {" "}

            amount : {props.expense.amount} {" "}
            createdAt : {props.expense.createdAt}  {" "}
            <Link to={`/edit/${props.expense.id}`}>Edit</Link>
        </p>
    </div>
    

);



export default expensetListItem;