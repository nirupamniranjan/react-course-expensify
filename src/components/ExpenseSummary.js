import React from 'react'
import addExpense from '../selectors/expenses-total'
import {connect} from 'react-redux'
import selectExpenses from '../selectors/expenses'


export class ExpenseSummary extends React.Component{
    
    render(){
        return (
            <div>
               <p>Viewing {this.props.expenses.length} expenses totalling {addExpense(this.props.expenses)}</p>
            </div>
        )
    }
}

const mapStateToProps =(state) =>{
    return{ 
    expenses:selectExpenses(state.expenses,state.filters)
    }
};

export default connect(mapStateToProps)(ExpenseSummary)