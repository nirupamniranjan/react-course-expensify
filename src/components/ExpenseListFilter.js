import React from 'react'
import {connect} from 'react-redux'
import {DateRangePicker} from 'react-dates'
import {setTextFilter,sortByAmount,sortByDate, setStartDate, setEndDate} from '../actions/filters'



export class ExpenseListFilters extends React.Component{

    state ={
        calanderFocused:null
    };

    onDatesChange=({startDate,endDate}) =>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange= (calanderFocused)=>{
        this.setState(()=>({calanderFocused}))
    }

    onTextChange = (e)=>{
        
        this.props.setTextFilter(e.target.value);
    };
    onSortChange =(e)=>{
        
        if(e.target.value==='date')
        this.props.sortByDate();
        else{
            this.props.sortByAmount()
        }
        };
render(){
   return (
    <div>
    <input type = "text" defaultValue ={this.props.filters.text} onChange={
        this.onTextChange
}
        
        />

        <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
            <option value ="date">
                Date
            </option>
            <option value="amount">
                Amount
            </option>
        </select>
        <DateRangePicker
         startDate={this.props.filters.startDate}
         startDateId={'startDate'}
         endDate={this.props.filters.endDate}
         endDateId={'endDateId'}
         onDatesChange={this.onDatesChange}
         focusedInput={this.state.calanderFocused}
         onFocusChange={this.onFocusChange}
         numberOfMonths={1}
         showClearDates={true}
         isOutsideRange={()=>{false}}
        
        />          

</div>

   );
    }

};
const mapStateToProps =(state) =>{
    return{ 
    filters:state.filters
    }
};

const mapDispatchToProps=(dispatch)=>({
   setTextFilter:(text)=> dispatch(setTextFilter(text)),
   sortByDate:()=>dispatch(sortByDate()),
   sortByAmount:()=>dispatch(sortByAmount()),
   setSortByDate:(startDate) =>dispatch(setStartDate(startDate)),
   setEndDate:(endDate) =>dispatch(setEndDate(endDate))

});

export default connect(mapStateToProps,mapStateToProps)(ExpenseListFilters)