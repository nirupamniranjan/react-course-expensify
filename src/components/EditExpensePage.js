import React from 'react';
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {editExpense,startRemoveExpense} from '../actions/expenses'


export class  EditExpensePage extends React.Component{
     onSubmit =(expense) =>{
        
        this.props.editExpense(this.props.expense.id,expense);
        this.props.history.push('/');
       

    };
    onClick =(e)=>{
        const id= this.props.expense.id
        this.props.startRemoveExpense({id})
        this.props.history.push('/')
    }
    
  render(){
    return (

    <div>
    <ExpenseForm expense ={this.props.expense} onSubmit={this.onSubmit}/>
    
    <button onClick={this.onClick}>Remove</button>
</div>
)}


};

const mapStateToProps =(state,props)=>{
    return {
        expense : state.expenses.find((expense)=>{
            return expense.id===props.match.params.id
        })

       
    
    }

};

const mapDispathToProps =(dispatch,props)=>({
    editExpense:(id,expense) => dispatch(addExpense(id,expense)),
    startRemoveExpense:(expense)=> dispatch(startRemoveExpense(expense))
});

export default connect(mapStateToProps,mapDispathToProps)(EditExpensePage);