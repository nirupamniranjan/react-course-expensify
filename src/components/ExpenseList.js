import React from 'react'
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
    <div>
       <h1>Expense List</h1>
        
       {

           props.expenses.lenght===0?(<p>No expenses</p>):(
           props.expenses.map((expenseItem)=>( <ExpenseListItem key={expenseItem.id} expense={expenseItem}/>)))
       }
       

    </div>
);

const mapStateToProps = (state)=>{
    return {
        expenses: selectExpenses(state.expenses,state.filters)
    };
  };
const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);



export default ConnectedExpenseList;