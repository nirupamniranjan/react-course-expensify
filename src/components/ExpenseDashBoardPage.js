
 import React from 'react';

 import ExpenseList from './ExpenseList'
 import ExpenseListFilter from './ExpenseListFilter';
 import ExpensesSummary from './ExpenseSummary'
const ExpenseDashboardPage = () =>(


<div>
<ExpenseListFilter />
    <ExpenseList/>
    <ExpensesSummary/>
</div>

);


export default ExpenseDashboardPage;



