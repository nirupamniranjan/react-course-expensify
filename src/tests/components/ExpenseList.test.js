import React from 'react'
import {shallow} from 'enzyme'
import {ExpenseList} from '../../components/ExpenseList'
import expenses from '../Fixtures/expenses'

test('should render ExpenseList with expenses',()=>{
    const wrapper = shallow(<ExpenseList expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});

test('Test should render ExpenseList with no expenses',()=>{
    const wrapper = shallow(<ExpenseList expenses={[]}/>);
    expect(wrapper).toMatchSnapshot();
});
