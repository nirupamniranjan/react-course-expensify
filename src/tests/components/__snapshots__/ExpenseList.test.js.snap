// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test should render ExpenseList with no expenses 1`] = `
<div>
  <h1>
    Expense List
  </h1>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <expensetListItem
    expense={
      Object {
        "amount": 125,
        "createdAt": 0,
        "description": "beer",
        "id": "1",
        "note": "",
      }
    }
    key="1"
  />
  <expensetListItem
    expense={
      Object {
        "amount": 12500,
        "createdAt": -259200000,
        "description": "Pizza",
        "id": "2",
        "note": "",
      }
    }
    key="2"
  />
  <expensetListItem
    expense={
      Object {
        "amount": 9000,
        "createdAt": 345600000,
        "description": "nuggets",
        "id": "3",
        "note": "",
      }
    }
    key="3"
  />
</div>
`;
